plugins/via-usbhub/fu-via-usbhub-device.c

```diff
@@ -0,0 +1,1275 @@
+					     G_IO_ERROR_FAILED,
+					     "failed to verify erase @0x%x: ", addr);
+				return FALSE;
+			}
+		}
+	}
+	return TRUE;
+}
+
+static gboolean
+fu_via_usbhub_device_write_block (FuViaUsbhubDevice *self,
+				  guint32 address,
+				  const guint8 *buf,
+				  gsize bufsz,
+				  GError **error)
+{
+	g_autofree guint8 *buf_tmp = g_malloc0 (bufsz);
+
+	/* sanity check */
+	if (bufsz > FU_VIA_USBHUB_TXSIZE) {
+		g_set_error (error,
+			     G_IO_ERROR,
+			     G_IO_ERROR_FAILED,
+			     "cannot write 0x%x in one block",
+			     (guint) bufsz);
+		return FALSE;
+	}
+
+	/* write */
+	if (!fu_via_usbhub_device_spi_write_enable (self, error)) {
+		g_prefix_error (error, "enabling SPI write failed: ");
+		return FALSE;
+	}
+	if (!fu_via_usbhub_device_spi_write_data (self, address, buf, bufsz, error)) {
+		g_prefix_error (error, "SPI data write failed: ");
+		return FALSE;
+	}
+	g_usleep (800);
+
+	/* verify */
+	if (!fu_via_usbhub_device_spi_read_data (self, address, buf_tmp, bufsz, error)) {
+		g_prefix_error (error, "SPI data read failed: ");
+		return FALSE;
+	}
+	return fu_common_bytes_compare_raw (buf, bufsz, buf_tmp, bufsz, error);
+}
+
+static gboolean
+fu_via_usbhub_device_write_blocks (FuViaUsbhubDevice *self,
+				   guint32 address,
+				   const guint8 *buf,
+				   gsize bufsz,
+				   GError **error)
+{
+	FuChunk *chk;
+	g_autoptr(GPtrArray) chunks = NULL;
+
+	/* write SPI data, then CRC bytes last */
+	chunks = fu_chunk_array_new (buf, bufsz, 0x0, 0x0, FU_VIA_USBHUB_TXSIZE);
+	if (chunks->len > 1) {
+		for (guint i = 1; i < chunks->len; i++) {
+			chk = g_ptr_array_index (chunks, i);
+			if (!fu_via_usbhub_device_write_block (self,
+								chk->address + address,
+								chk->data,
+								chk->data_sz,
+								error)) {
+				g_prefix_error (error, "failed to write block 0x%x: ", chk->idx);
+				return FALSE;
+			}
+			fu_device_set_progress_full (FU_DEVICE (self),
+						     (gsize) i - 1,
+						     (gsize) chunks->len);
+		}
+	}
+	chk = g_ptr_array_index (chunks, 0);
+	if (!fu_via_usbhub_device_write_block (self,
+						chk->address + address,
+						chk->data,
+						chk->data_sz,
+						error)) {
+		g_prefix_error (error, "failed to write CRC block: ");
+		return FALSE;
+	}
+	fu_device_set_progress_full (FU_DEVICE (self), (gsize) chunks->len, (gsize) chunks->len);
+	return TRUE;
+}
+
+static gboolean
+fu_via_usbhub_device_update_v1 (FuViaUsbhubDevice *self,
+				GBytes *fw,
+				GError **error)
+{
+	gsize bufsz = 0;
+	const guint8 *buf = g_bytes_get_data (fw, &bufsz);
+
+	/* erase */
+	fu_device_set_status (FU_DEVICE (self), FWUPD_STATUS_DEVICE_ERASE);
+	if (!fu_via_usbhub_device_erase_all (self, error)) {
+		g_prefix_error (error, "failed to erase chip: ");
+		return FALSE;
+	}
+
+	/* write in chunks */
+	if (!fu_via_usbhub_device_write_blocks (self, 0x0, buf, bufsz, error))
+		return FALSE;
+
+	/* success */
+	fu_device_set_status (FU_DEVICE (self), FWUPD_STATUS_DEVICE_RESTART);
+//FIXME: replug?
```

Comment:
what exactly happens while in this case?  does the device restart on it's own after all the chunks are done?

Reply:
At the moment the device does nothing and remains in the old firmware version until the user manually replugs. Apparently there's a reboot command we can use, and I'm just waiting for details on whether it's supported by V2 or V? protocols.

Reply:
I guess the ones that don't use reboots could probably have the usable while flashing flag set too?

Reply:
I've added the reboot now, and apparently the device is not fully functional when flashing.